### PR TITLE
(2/n) Data Refactor - Data Modules

### PR DESCRIPTION
**Pull request by awaelchli**
Adds implementations of data modules and tests for all finetuning datasets we support. 

A separate PR will integrate them into the scripts to keep the scope of the PR low and iterate faster. Not removing the existing `scripts/prepare_*.py` to not break the main branch during the refactor.

Follow-up PRs:
- Integration into finetuning scripts (progress in #950)
- Remove old implementations under scripts/ and tests/
- Implement disk cache for medium size datasets (e.g. FLAN)
- Update the markdown doc files

Note: If we look at the data modules right now, they look like a lot of boilerplate code repeated. There is probably a way to make a generic data module that can load jsons and do everything. There are however tiny differences between the datasets and parameters that I found it easier to implement individual modules even if some boilerplate code is repeated.

### CODE REVIEW

File: lit_gpt/data/base.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+class SFTDataset(Dataset):
+    """An in-memory dataset for supervised finetuning with `input_ids` and `labels`.
+
+    Args:
+        data: A list of samples (dicts). The target/label must be stored under the key 'output' and the instruction
+            or other data can be stored under any key as long as it is compatible with the given prompt template.
+        tokenizer: The tokenizer to use. Should match the one that was used to pretrain the model.
+        prompt_template: A prompt template (format string or callable)
+        max_seq_length: Truncate sequences that are longer than this value. By default, no truncation is applied.
+        mask_prompt: Whether to mask the prompt section from the label (with ``ignore_index``)
```
Comment (carmocca) [inline lit_gpt/data/base.py:42]:
```suggestion
        prompt_template: A prompt template (format string or callable).
        max_seq_length: Truncate sequences that are longer than this value. By default, no truncation is applied.
        mask_prompt: Whether to mask the prompt section from the label (with ``ignore_index``).
```

File: tests/data/conftest.py
Comment (carmocca) [inline tests/data/conftest.py:1]: Cool idea

**Review by carmocca — APPROVED** [review]

File: lit_gpt/data/data_module.py
[the remark this answers is not in the document]
Comment (awaelchli) [reply lit_gpt/data/data_module.py:1]: Yes because it no longer only contains just the data module but also other base classes. Do you have suggestions or would you keep the data_module name anyway?

[the remark this answers is not in the document]
Comment (carmocca) [reply lit_gpt/data/data_module.py:1]: No strong opinion. Looks fine if you prefer it this way

File: lit_gpt/data/base.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    batched = {}
+    for key in ("input_ids", "labels"):
+        pad_value = pad_id if key == "input_ids" else ignore_index
+
+        # Pad right based on the longest sequence
+        batched[key] = torch.stack([
+            torch.nn.functional.pad(sample[key], (0, longest - len(sample[key])), value=pad_value)
+            for sample in samples
+        ])
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply lit_gpt/data/base.py]: Great, this works. Eliminates the need to compute the longest element and padding size.
